tools/rpkg/src/statement.cpp

```diff
@@ -20,66 +21,35 @@ static void VectorToR(Vector &src_vec, size_t count, void *dest, uint64_t dest_o
 	}
 }
 
-SEXP RApi::Release(SEXP stmtsexp) {
-	if (TYPEOF(stmtsexp) != EXTPTRSXP) {
-		cpp11::stop("duckdb_release_R: Need external pointer parameter");
+[[cpp11::register]] void rapi_release(duckdb::stmt_eptr_t stmt) {
+	auto stmt_ptr = stmt.release();
+	if (stmt_ptr) {
+		delete stmt_ptr;
 	}
-	RStatement *stmtholder = (RStatement *)R_ExternalPtrAddr(stmtsexp);
-	if (stmtsexp) {
-		R_ClearExternalPtr(stmtsexp);
-		delete stmtholder;
-	}
-	return R_NilValue;
-}
-
-static SEXP duckdb_finalize_statement_R(SEXP stmtsexp) {
-	return RApi::Release(stmtsexp);
 }
 
-SEXP RApi::Prepare(SEXP connsexp, SEXP querysexp) {
-	RProtector r;
-	if (TYPEOF(querysexp) != STRSXP || Rf_length(querysexp) != 1) {
-		cpp11::stop("duckdb_prepare_R: Need single string parameter for query");
-	}
-	if (TYPEOF(connsexp) != EXTPTRSXP) {
-		cpp11::stop("duckdb_prepare_R: Need external pointer parameter for connections");
-	}
-
-	char *query = (char *)CHAR(STRING_ELT(querysexp, 0));
-	if (!query) {
-		cpp11::stop("duckdb_prepare_R: No query");
-	}
-
-	auto conn_wrapper = (ConnWrapper *)R_ExternalPtrAddr(connsexp);
-	if (!conn_wrapper || !conn_wrapper->conn) {
-		cpp11::stop("duckdb_prepare_R: Invalid connection");
+[[cpp11::register]] cpp11::list rapi_prepare(duckdb::conn_eptr_t conn, std::string query) {
+	if (!conn || !conn->conn) {
+		cpp11::stop("rapi_prepare: Invalid connection");
 	}
 
-	auto stmt = conn_wrapper->conn->Prepare(query);
+	auto stmt = conn->conn->Prepare(query.c_str());
```

Comment:
How about implementing multi-statement preparation (#1102) also for the R-client? I had just got it working with the old version of rapi (because this functionality will be needed to bundle option changing `PRAGMA`s to lazy queries), but then noticed that there will be substantial changes in this PR, so no idea to send a separate overlapping PR before this gets merged.

Certainly it would be straightforward to add here together with all other changes. By copying as much as possible from the [Python version](https://github.com/duckdb/duckdb/blob/b845e1d7ee989bc9c9bf1c724c440e72d92bbdbe/tools/pythonpkg/src/pyconnection.cpp#L114-L128), it would be something like the following, i.e. replacing the line 36 with:
```cpp
	auto statements = conn->conn->ExtractStatements(query.c_str());
	if (statements.empty()) {
	  // no statements to execute
	  cpp11::stop("rapi_prepare_R: No statements to execute");
	}
	// if there are multiple statements, we directly execute the statements besides the last one
	// we only return the result of the last statement to the user, unless one of the previous statements fails
	for (idx_t i = 0; i + 1 < statements.size(); i++) {
	  auto res = conn->conn->Query(move(statements[i]));
	  if (!res->success) {
	    cpp11::stop("rapi_prepare_R: Failed to execute statement %s\nError: %s", query.c_str(), res->error.c_str());
	  }
	}
	auto stmt = conn->conn->Prepare(move(statements.back()));
```
@Mytherin and @hannesmuehleisen, what do you think?

Reply:
@rsund sure lets add this, but let's do a separate PR